### PR TITLE
add query for built-in actors; fix eth_getCode and eth_getStorageAt.

### LINKED ISSUE #506: `eth_getCode` and `eth_getStorageAt` unable to deal with non-EVM actors
These methods currently return errors if they're called on non-EVM actors (e.g. accounts or else). Instead, they should return empty success responses. See Lotus:

https://github.com/filecoin-project/lotus/blob/b0bc4a96320942c0673a3cab08314074edf4b4f9/node/impl/full/eth.go#L450

### PR DESCRIPTION
**Pull request by raulk**
Aims to fix #506 in a more stable way than #511. Needs tests and cleanup.

### CODE REVIEW

File: fendermint/fendermint/eth/api/src/state.rs [hunk 1]
```diff
@@ -388,6 +401,30 @@ where
...
+        let Some((
+            _,
+            ActorState {
+                code: actor_type_cid,
+                ..
+            },
+        )) = self.client.actor_state(&addr, height).await?.value
+        else {
+            return Ok(ActorType::Inexistent);
+        };
```
Comment (raulk) [inline fendermint/fendermint/eth/api/src/state.rs:420]: Thanks rustfmt for some ugly formatting here...

File: fendermint/fendermint/vm/interpreter/src/fvm/state/query.rs [hunk 2]
```diff
@@ -204,6 +205,25 @@ where
         &self.state_params
     }
 
+    /// Returns the registry of built-in actors as enrolled in the System actor.
+    pub async fn builtin_actors(self) -> anyhow::Result<(Self, Vec<(String, Cid)>)> {
```
Comment (raulk) [inline fendermint/fendermint/vm/interpreter/src/fvm/state/query.rs:209]: Eventually we may want to add a `debug_ipcBuiltinActors` JSON-RPC operation to return known built-in actors for debugging.

File: fendermint/fendermint/vm/interpreter/src/fvm/state/query.rs [hunk 3]
```diff
@@ -7,13 +7,14 @@ use std::{cell::RefCell, sync::Arc};
 use anyhow::{anyhow, Context};
 
 use cid::Cid;
-use fendermint_vm_actor_interface::system::is_system_addr;
+use fendermint_vm_actor_interface::system::{is_system_addr, State, SYSTEM_ACTOR_ADDR};
```
Comment (aakoshh) [inline fendermint/fendermint/vm/interpreter/src/fvm/state/query.rs:10]:
```suggestion
use fendermint_vm_actor_interface::system::{is_system_addr, State => SystemState, SYSTEM_ACTOR_ADDR};
```

I got a bit confused what `State` we were referring to.

File: fendermint/fendermint/vm/interpreter/src/fvm/state/query.rs [hunk 4]
```diff
@@ -204,6 +205,25 @@ where
         &self.state_params
     }
 
+    /// Returns the registry of built-in actors as enrolled in the System actor.
+    pub async fn builtin_actors(self) -> anyhow::Result<(Self, Vec<(String, Cid)>)> {
+        let (s, sys_state) = {
+            let (s, state) = self.actor_state(&SYSTEM_ACTOR_ADDR).await?;
+            (s, state.ok_or(anyhow!("no system actor"))?.1)
+        };
+        let state: State = s
```
Comment (aakoshh) [inline fendermint/fendermint/vm/interpreter/src/fvm/state/query.rs:214]:
```suggestion
        let state: SystemState = s
```

File: fendermint/fendermint/vm/interpreter/src/fvm/state/query.rs [hunk 5]
```diff
@@ -204,6 +205,25 @@ where
...
+            .store
+            .get_cbor(&sys_state.code)
```
Comment (aakoshh) [inline fendermint/fendermint/vm/interpreter/src/fvm/state/query.rs:216]:
```suggestion
            .get_cbor(&sys_state.state)
```

I think you're getting the wrong piece of data there. The [system state](https://github.com/consensus-shipyard/ipc-monorepo/blob/03b745c7b64c2cb7888fbffd75d36218f02c02ae/fendermint/fendermint/vm/interpreter/src/fvm/genesis.rs#L151-L153) is indeed the manifest but it's [not stored in the code field](https://github.com/consensus-shipyard/ipc-monorepo/blob/03b745c7b64c2cb7888fbffd75d36218f02c02ae/fendermint/fendermint/vm/interpreter/src/fvm/state/genesis.rs#L171-L181). 

It looks like a typo but got me confused if there is a trick I forgot. This is where using `TCid` would be useful.

**Review by aakoshh — COMMENTED** [review]
Nice, you touched the entire stack of query handling :trophy:  
I hope it wasn't too bad of an experience, there are quite a few pieces.

**Comment by aakoshh** [comment]
Could you cherry pick these tests? 
* https://github.com/consensus-shipyard/ipc-monorepo/pull/511/commits/8ac32ecc17f520cb331e21f117d53475e6dd0e8c
* https://github.com/consensus-shipyard/ipc-monorepo/pull/511/commits/a3eec2669d1124e5125b9de73af6a5dc951b11d3

They don't do much but would catch if we returned something `ethers` cannot parse.

**Comment by raulk** [comment]
Continued and merged in https://github.com/consensus-shipyard/ipc/pull/610.

**Pull request closed without merging**